### PR TITLE
add Hazelcast session management #1571

### PR DESCRIPTION
**Pull request by olamy**
still WIP. I need to finish module etc...
Just for testing build on our Jenkins.

### CODE REVIEW

**Comment by olamy** [comment]
@janbartel please review.
Documentation will be part of an other pr.

**Comment by olamy** [comment]
@WalkerWatch documentation page added (very similar to Infinispan)

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    private String jettySessionMapName;
+
+    private IMap<String, SessionData> sessionDataMap;
+
+    public HazelcastSessionDataStore()
+    {
+        // no op
+    }
+
+    public HazelcastSessionDataStore( HazelcastInstance hazelcastInstance, String jettySessionMapName )
```
Comment (janbartel) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: Are you sure you want to have both constructor params and getters and setters for these? I thought usually one had one or the other, not both - in which case I'd go for the setters instead.

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    public void setJettySessionMapName( String jettySessionMapName )
+    {
+        this.jettySessionMapName = jettySessionMapName;
+    }
+
+
+    @Override
+    public SessionData load( String id )
+        throws Exception
+    {
```
Comment (janbartel) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: Take a look at the load method InfinispanSessionDataStore - we ensure we run it in the scope of a the webapp's classloader with _context.run(Runnable), and we also use atomics to make sure that 2 or more simultaneous requests only result in a single load.

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        _context = context;
+        if (this.sessionDataMap == null)
+        {
+            this.sessionDataMap = getHazelcastInstance().getMap( getJettySessionMapName() );
+        }
+    }
+
+    @Override
+    public void store( String id, SessionData data )
```
Comment (janbartel) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: Why do you need to override the store method? Implementing doStore() should be enough.

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    public void doStore( String id, SessionData data, long lastSaveTime )
+        throws Exception
+    {
+        this.sessionDataMap.put( id, data);
+    }
+
+    @Override
+    public boolean isPassivating()
+    {
+        return false;
```
Comment (janbartel) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: I don't know anything about hazelcast, but are you sure it doesn't ever serialize the session data? If it does, then the correct response here is "true".

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        throws Exception
+    {
+        this.sessionDataMap.put( id, data );
+    }
+
+    @Override
+    public void doStore( String id, SessionData data, long lastSaveTime )
+        throws Exception
+    {
+        this.sessionDataMap.put( id, data);
```
Comment (janbartel) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: Multiple contexts can share the same session id, but they are not allowed to share content. Thus, we usually make a unique key by combining unique information about the context with the session id, eg:  context_path+first virtual host+session id.  See InfinispanSessionDataStore.getCacheKey() method.  Or have you configured a separate hazelcast map for each and every context?

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+            }
+            return false;
+        } ).collect( Collectors.toSet() );
+    }
+
+    @Override
+    public boolean exists( String id )
+        throws Exception
+    {
+        return this.sessionDataMap.containsKey( id );
```
Comment (janbartel) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: Is there a performance difference in hazelcast between operations over the collection like containsKey() vs get()? Eg I know there is in infinispan and a get() op is actually cheaper.

File: tests/test-sessions/test-hazelcast-sessions/src/test/java/org/eclipse/jetty/hazelcast/session/ClusteredLastAccessTimeTest.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.hazelcast.session;
+
+import org.eclipse.jetty.server.session.AbstractClusteredLastAccessTimeTest;
+import org.eclipse.jetty.server.session.SessionDataStoreFactory;
+
+public class ClusteredLastAccessTimeTest
```
Comment (janbartel) [inline tests/test-sessions/test-hazelcast-sessions/src/test/java/org/eclipse/jetty/hazelcast/session/ClusteredLastAccessTimeTest.java]: I guess these tests use  embedded hazelcast. Is there any point in also checking with a "remote" hazelcast instance (eg for serialization problems)?

File: jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+--module=session-store-hazelcast-remote
+
+#jetty.session.hazelcast.jettySessionMapName=jetty_sessions
+#jetty.session.hazelcast.onlyClient=true
+#jetty.session.hazelcast.configurationLocation=
+#jetty.session.gracePeriod.seconds=3600
+#jetty.session.savePeriod.seconds=0
+----
+
+jetty.session.hazelcast.jettySessionMapName::
```
Comment (janbartel) [inline jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc]: Trivial nitpick:  is there any point in repeating jettySession in the name of this property when it is prefixed by jetty.session?

File: jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+[source, screen, subs="{sub-order}"]
+----
+# ---------------------------------------
+# Module: session-store-hazelcast-remote
+# Enables session data store in a remote Hazelcast Map
+# ---------------------------------------
+--module=session-store-hazelcast-remote
+
+#jetty.session.hazelcast.jettySessionMapName=jetty_sessions
+#jetty.session.hazelcast.onlyClient=true
```
Comment (janbartel) [inline jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc]: jetty.session.hazelcast.onlyClient doesn't match the name of the property in the doc.

File: jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc (hunk 8 above)
Comment (WalkerWatch) [inline jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc]: Can you please make sure to add this .adoc to the[ chapter.adoc](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-documentation/src/main/asciidoc/administration/sessions/chapter.adoc) so it is included when the documentation is built?

Comment (olamy) [inline jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc]: fix see #1584

**Comment by olamy** [comment]
Infinispan fix revert and now part of #1584

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java (hunk 6 above)
Comment (olamy) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: containsKey is faster (confirmed by Christoph Engelbert Hazelcast committer https://twitter.com/olamy/status/870134676769316864 )

File: tests/test-sessions/test-hazelcast-sessions/src/test/java/org/eclipse/jetty/hazelcast/session/ClusteredLastAccessTimeTest.java (hunk 7 above)
Comment (olamy) [inline tests/test-sessions/test-hazelcast-sessions/src/test/java/org/eclipse/jetty/hazelcast/session/ClusteredLastAccessTimeTest.java]: well just to be sure we connect as a client and serialize/deserialize correctly

File: jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc (hunk 8 above)
Comment (olamy) [inline jetty-documentation/src/main/asciidoc/administration/sessions/session-configuration-hazelcast.adoc]: sounds look but with camel case :-)

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java (hunk 2 above)
Comment (olamy) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: thanks for the tip on classloader!
done.

**Comment by olamy** [comment]
Thanks for review guys.
Changes done (please review again :-) )

File: tests/test-sessions/test-hazelcast-sessions/pom.xml [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+      <artifactId>jetty-jmx</artifactId>
+      <version>${project.version}</version>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty.toolchain</groupId>
+      <artifactId>jetty-test-helper</artifactId>
+      <scope>test</scope>
+    </dependency>
+
```
Comment (janbartel) [inline tests/test-sessions/test-hazelcast-sessions/pom.xml]: Do these tests require hazelcast to be locally installed? If so, they need to be disabled by default, and a profile used to selectively enable with a property - see the infinispan/mongo test sessions poms for examples.

Comment (olamy) [inline tests/test-sessions/test-hazelcast-sessions/pom.xml]: Nope. client tests connect "remotely" to a started Hazelcast instance running in the same vm

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+            try
+            {
+                SessionData sd = load(candidate);
+
+                //if the session no longer exists
+                if (sd == null)
+                {
+                    if (LOG.isDebugEnabled())
+                    {
+                        LOG.debug( "Session {} does not exist in infinispan", candidate );
```
Comment (emre-aydin) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: I guess this should change to Hazelcast :)

Comment (olamy) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: @emre-aydin thanks typo should be fixed now

**Comment by olamy** [comment]
on hold see #1585

**Comment by olamy** [comment]
expect review from hazelcast @emre-aydin

File: jetty-hazelcast/src/main/config/modules/session-store-hazelcast-embedded.mod [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+[provides]
+session-store
+
+[depend]
+sessions
+
+[files]
+maven://com.hazelcast/hazelcast/3.8.2|lib/hazelcast/hazelcast-3.8.2.jar
+maven://com.hazelcast/hazelcast-client/3.8.2|lib/hazelcast/hazelcast-client-3.8.2.jar
```
Comment (mesutcelik) [inline jetty-hazelcast/src/main/config/modules/session-store-hazelcast-embedded.mod]: you normally should not need to include client jar in embedded mode.

File: jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        throws Exception
+    {
+        _context = context;
+    }
+
+    @Override
+    public void doStore( String id, SessionData data, long lastSaveTime )
+        throws Exception
+    {
+        this.sessionDataMap.put( getCacheKey( id ), data);
```
Comment (mesutcelik) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: We normally prefer`.set` if the old value is not needed. 
https://github.com/hazelcast/hazelcast/blob/v3.8.2/hazelcast/src/main/java/com/hazelcast/core/IMap.java#L790

Comment (mesutcelik) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: The other approach would be to update only dirty attributes via Hazelcast Entry Processor. I see you have a session wide [dirty flag](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionData.java#L57) but not attribute based one. It looks like someone started to implement it but [not finished yet](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionData.java#L163) assuming the argument `name` is attribute name.

You can see [Entry Processor based approach](https://github.com/hazelcast/hazelcast-wm/blob/v3.8.1/src/main/java/com/hazelcast/web/entryprocessor/SessionUpdateEntryProcessor.java#L78) in our ServletFilter based Session Replication Module. We don't have this feature in Hazelcast Tomcat/Jetty Native Implementations.

Comment (olamy) [inline jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java]: fixed for put vs set.
 Thanks!

**Comment by olamy** [comment]
@janbartel so I think you can merge it.

**Pull request merged**